#167190527 JWT Authentication

JuwonAbiola opened the pull request:
#### What does this PR do?
- Generates token 
- Verifies token 

#### Description of Task proposed in this pull request?
- Install jsonwebtoken NPM package 
- Create the generate token function that accepts payload and expiration time
- Create the verify token function that accepts token

#### How should this be manually tested (Quality Assurance)?
- check out to this branch ch-create-token-167190527
- run npm install to download dependencies
- Add JWT_KEY with value to the .env file 
- run npm test

#### What are the relevant pivotal tracker stories?
- [#167190527](https://www.pivotaltracker.com/story/show/167190527)

#### Any background context you want to add (Operations Impact)?
- After cloning this repo 
- Checkout to this branch ch-create-token-167190527
- Run npm install
- Add JWT_KEY to the .env file 

#### What I have learned working on this feature: 
- I learned how to do unit testing 

#### Screenshots: N/A

Adekoreday reviewed (commented):
great job, kindly factor a case of passing integers as parameters to generateToken function in your test

JuwonAbiola commented:
> great job, kindly factor a case of passing integers as parameters to generateToken function in your test

Thanks for the feedback I will factor passing numbers as parameters

Adekoreday reviewed (dismissed):
great job

chialuka commented on test/helpers/generateToken.test.js:
```diff
@@ -0,0 +1,23 @@
+describe('Test generateToken function', () => {
+  it('should return a token', async () => {
+    const payload = { id: 1 };
+    const token = await generateToken(payload, '24h');
```
Since in the helpers this function for `generateToken` isn't asynchronous, is there any need to await it here in the test?

ayodejiAA reviewed (dismissed):
Well done

dinobi commented on package.json:
```diff
@@ -6,7 +6,7 @@
     "lint": "eslint '**/*.js'",
     "lint:fix": "prettier-eslint '**/*.js' --write",
-    "test:staged": "nyc mocha --timeout 1000 -r esm --exit && nyc check-coverage --statements 100 --functions 100 --lines 100 --branches 100",
+    "test:staged": "npm test && nyc check-coverage --statements 100 --functions 100 --lines 100 --branches 100",
```
Why do we have this change?

JuwonAbiola replied:
I changed it because it is the same as the test script

dinobi commented on package.json:
```diff
@@ -16,7 +16,7 @@
     "start": "node dist/index.js",
     "start:dev": "cross-env NODE_ENV=development DEBUG=dev nodemon --exec babel-node server",
-    "test": "nyc mocha --timeout 1000 -r esm --exit"
+    "test": "nyc mocha ./test/index.test.js --timeout 1000 -r esm --exit"
```
ditto

JuwonAbiola replied:
Since the team agreed on having an index file for every folder I decided to import all test files into the index file and make the test run from the index file

dinobi commented on package.json:
```diff
@@ -30,14 +30,15 @@
-    "sqlite3": "^4.0.9"
+    "sqlite3": "^4.0.9",
+    "swagger-ui-express": "^4.0.7",
+    "yamljs": "^0.3.0"
```
ditto

JuwonAbiola replied:
This is as a result of the husky configuration to format the code before pushing

dinobi commented on server/helpers/generateToken.js:
```diff
@@ -0,0 +1,16 @@
+ * @param {string} expiresIn
+ * @return {string} token
+ */
+const generateToken = (payload, expiresIn) => (!expiresIn
```
Is it not better to have a default expiresIn and remove the ternary operation?

dinobi replied:
aren't we also have the expires in set in our environment?

JuwonAbiola replied:
The purpose of not setting the expiresIn in environment variables is simply because the expiration time may vary for different occasions moving on I'll set a default expiresIn and set  it in environment

dinobi commented on test/helpers/generateToken.test.js:
```diff
@@ -0,0 +1,28 @@
+import chai from 'chai';
+import helpers from '../../server/helpers';
```
please update your test to show that, when an expiresIn is not provided, we use a default value

dinobi commented on test/helpers/generateToken.test.js:
```diff
@@ -0,0 +1,28 @@
+
+const { expect } = chai;
+
+describe('Test generateToken function', async () => {
```
Please avoid `async` if we are not awaiting anything

dinobi commented on test/root.test.js:
```diff
@@ -0,0 +1,21 @@
+import chai from 'chai';
```
What is `root.test.js`?, can you find a way to structure your test to the files that they are testing

dinobi reviewed (changes requested):
I'm thinking you need to also create a method a helper method that verifies a token for our protected endpoint, please read the acceptance criteria for this story very carefully. So please change the name of the PR if possible.

dinobi commented on test/middlewares/verifyToken.test.js:
```diff
@@ -0,0 +1,83 @@
+describe(
+  'verify token middleware', () => {
+    afterEach(() => {
+      if (jwt.verify.restore) jwt.verify.restore();
```
@jhayX  Why are you stubbing out jwt?

dinobi reviewed (commented):
Please update your test to use the session creation and generated token

dinobi commented on test/middlewares/verifyToken.test.js:
```diff
@@ -0,0 +1,84 @@
+  }
+);
+
+export default chai;
```
Why are you exporting chai

JuwonAbiola replied:
In order for the test file to be accessible on the index file, it is required

dinobi commented on test/users/signup.test.js:
```diff
@@ -91,3 +91,5 @@ describe('POST User', () => {
   });
 });
+
+export default chai;
```
what are you exporting `chai`?

JuwonAbiola committed:
feature-[167190527]:generate token
- create generate token function
- test generate token function
[Delivers #167190527]
